### ISSUE TITLE
Debugger halts but shows no information. StackFrames are empty

### ISSUE THREAD

**Issue by keckelt** [microsoft/ptvsd#1979]
## Environment data

- PTVSD version: 4.3.2
- OS and version: Win10 19002 (Insiders) with  Ubuntu 18.04.3 LTS as Linux Subsytem (WSL2)
- Python version: 3.7 (in python:3.7 docker container)
- Using VS Code or Visual Studio: VS Code 1.40.2 (Python 2019.11.50794 as remote extension)

My code is inside the WSL, located at `~/cht`.
The directory is mounted as `/phovea` in the docker container:
```
    volumes:
      - '.:/phovea'
```
The port 5678 is mapped to 5678.
I'm running it in the Docker WSL2 preview.

This is my config in launch.json:
```json
    {
      "name": "Python: phovea_server",
      "type": "python",
      "request": "attach",
      "logToFile": true,
      "port": 5678,
      "host": "localhost",
      "pathMappings": [
        {
          "localRoot": "${workspaceFolder}",
          "remoteRoot": "/phovea"
        }
      ]
    },
```

## Actual behavior

I can attach the debugger and execution also breaks on breakpoints.
However, 
* the variables pane is empty
* I don't see where the execution halted (
* I can not _Debug: Evaluate_ (prints: "Unable to find thread for evaluation.")
* Watched variables are _not available_ 

I already read https://github.com/microsoft/ptvsd/issues/1006 and https://github.com/Microsoft/ptvsd/issues/1325.
Here my debug.log: https://gist.github.com/keckelt/1470e1f5ba9ae67df1414eb57c3739be
As you can see, the StackFrames array is empty.

## Expected  behavior

See variables and where execution was paused.

## Steps to reproduce:
Please see above

**Comment by fabioz** [comment]
Do you know if phovea uses some async framework? (such as gevent -- the fact that it cannot find a thread for evaluation hints at that)

**Comment by keckelt** [comment]
yes it does.

**Comment by fabioz** [comment]
I think that `ptvsd` 4.x does not really support gevent debugging properly, you have to upgrade to `ptvsd 5.x` and set an environment variable `GEVENT_SUPPORT=True`.

@int19h can users already use `ptvsd 5.x`? Is it available in the insiders version?

**Comment by int19h** [comment]
It's selectively enabled, but we don't have a way to explicitly opt into it just yet (it's not as simple as installing the new version, because the extension had to be changed to accommodate it, as well - if it thinks it's dealing with the old debugger, it won't do the right things for the new one).

@keckelt, I think the error you're getting is just an editor warning complaining about mismatched JSON schema. This is wrong, since the property is actually there, but it shouldn't prevent you from starting the debugger.

**Comment by keckelt** [comment]
Thanks!
I'll give it a try today.

**Comment by keckelt** [comment]
I installed VS Code Insiders today (1.42):
```
Version: 1.42.0-insider (system setup)
Commit: 22a87bf3f259d3fcff66f636a9db611741c7e1c5
Date: 2019-12-17T05:41:30.616Z
Electron: 6.1.6
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.19041
```

I also switched to the insiders version of the Python extension, which is now Version: 2020.1.54965-dev
settings.json: ` "python.insidersChannel": "daily"`

Still having the same errors unfortunately.

[debug5101.log](https://github.com/microsoft/ptvsd/files/3972553/debug5101.log)

**Comment by keckelt** [comment]
Wait ... 🤔 
I was totally focused on VS Code. Should i also use/install the PTVSD v5 pre-release in my app? https://pypi.org/project/ptvsd/#history

**Comment by int19h** [comment]
If you're using attach by TCP, then yes, it's the version of ptvsd that you're explicitly importing or using to launch your code that matters.

**Comment by int19h** [comment]
Did you get better luck with ptvsd 5?

**Comment by keckelt** [comment]
Sorry there were a few other things coming up after the holidays and I didn't have time to test it yet.
Will give it another shot on Friday.

**Comment by keckelt** [comment]
I've updated my Code Insider to this version:
```
Version: 1.42.0-insider (system setup)
Commit: f3dbcea32a6b52ce819dc863e778fb8517357f61
Date: 2020-01-24T05:38:14.538Z
Electron: 7.1.7
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.19041
```

Also the VS Code Server for WSL:
```
Updating VS Code Server to version f3dbcea32a6b52ce819dc863e778fb8517357f61
```


And the Python extension:
```
Name: Python
Id: ms-python.python
Description: Linting, Debugging (multi-threaded, remote), Intellisense, Jupyter Notebooks, code formatting, refactoring, unit tests, snippets, and more.
Version: 2020.1.58038
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-python.python
```

I also installed the most recent ptvsd release [v5.0.0a12](https://github.com/microsoft/ptvsd/releases/tag/v5.0.0a12).

My config still looks like this:
```
{
  "name": "Python: phovea_server",
  "type": "python",
  "request": "attach",
  "logToFile": true,
  "port": 5678,
  "host": "localhost",
  "gevent": true,
  "pathMappings": [
    {
      "localRoot": "${workspaceFolder}",
      "remoteRoot": "/phovea"
    }
  ]
}
```

Additionally i've set an environment variable with docker compose:
```
version: '2.0'
services:
  api:
    ports:
      - '5678:5678'
    environment:
      GEVENT_SUPPORT: 'true'
```


---

Then i couldn't attach to the debugger due to the error `Invalid message: Session is not initialized yet`

I saw https://github.com/microsoft/ptvsd/issues/1854#issuecomment-553235992 and added `"debugServer": 5678,` to my launch.json, i.e.:

```
{
  "name": "Python: phovea_server",
  "type": "python",
  "request": "attach",
  "logToFile": true,
  "port": 5678,
  "debugServer": 5678,
  "host": "localhost",
  "pathMappings": [
    {
      "localRoot": "${workspaceFolder}",
      "remoteRoot": "/phovea"
    }
  ]
}
```

---

After being able to attach, i now get the following error when a breakpoint should be hit:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/gevent/pywsgi.py", line 976, in handle_one_response
    self.run_application()
  File "/usr/local/lib/python3.7/site-packages/geventwebsocket/handler.py", line 87, in run_application
    return super(WebSocketHandler, self).run_application()
  File "/usr/local/lib/python3.7/site-packages/gevent/pywsgi.py", line 923, in run_application
    self.result = self.application(self.environ, self.start_response)
  File "/usr/local/lib/python3.7/site-packages/werkzeug/middleware/proxy_fix.py", line 232, in __call__
    return self.app(environ, start_response)
  File "/phovea/phovea_server/phovea_server/dispatcher.py", line 62, in __call__
    return app(environ, start_response)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2463, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2449, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1866, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python3.7/site-packages/flask_login/utils.py", line 261, in decorated_view
    return func(*args, **kwargs)
  File "/phovea/my/test/script.py", line 44, in test_method   <--------------------------------------  breakpoint here
    try:
  File "/usr/local/lib/python3.7/site-packages/ptvsd/_vendored/pydevd/_pydevd_frame_eval/pydevd_frame_tracing.py", line 36, in _pydev_stop_at_break
    if t.additional_info.is_tracing:
AttributeError: '_DummyThread' object has no attribute 'additional_info'
2020-01-24T15:29:48Z {'REMOTE_ADDR': '172.18.0.1', 'REMOTE_PORT': '60800', 'HTTP_HOST': 'localhost:8080', (hidden keys: 30)} failed with AttributeError
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/gevent/pywsgi.py", line 976, in handle_one_response
    self.run_application()
  File "/usr/local/lib/python3.7/site-packages/geventwebsocket/handler.py", line 87, in run_application
    return super(WebSocketHandler, self).run_application()
```

I could resolve that by disabling the use of werkzeug's relaoder.


Now the debugger stops/resumes without errors but i still don't see a stacktrace/variables when stopping in one of my flask routes:
![image](https://user-images.githubusercontent.com/10337788/73082739-4af35b80-3eca-11ea-9121-d6cd350e9bd5.png)


I hope this debug log is the right one ( it's a different filename schema than before):
[debugger.vscode_3c97d786-dcbe-42c9-b92c-62212af93ffc.log](https://github.com/microsoft/ptvsd/files/4109361/debugger.vscode_3c97d786-dcbe-42c9-b92c-62212af93ffc.log)

**Comment by int19h** [comment]
For debug logs from the container, set `PTVSD_LOG_DIR` environment variable - this should produce a bunch of `ptvsd.*.log` files in the directory specified.

To get around the need to use "debugServer", install the Insiders build for the extension: 
https://github.com/microsoft/ptvsd/issues/1706#issuecomment-572287267

but that will most likely get you to the error about the thread...

@fabioz, it looks like something is also trying to muck around with the Python thread objects here? and presumably that's not gevent, since the flag for that is already enabled.

**Comment by keckelt** [comment]
I installed the linked vsix, and the python extension now says `Version: 2020.2.60336-dev`.
I also added the two entries to my settings.json. 

I'm not sure why but it also shows: `"python.insidersChannel": "off"`  -  should I reset that to daily again?

---

I still have the same error now -  one thing i notices though:
When the debugger halts at the breakpoint the mainthread and dummy threads are paused.
I continue the main thread and the others stay paused and I have to continue one so that the mainthread steps:
![ptvsd_debugging](https://user-images.githubusercontent.com/10337788/73248904-188c7b80-41b4-11ea-9eda-a0792ee144ea.gif)
